src/components/builder/i18n.tsx

```diff
@@ -1,173 +1,159 @@
-import {SupportedLocales, TranslationsContainer} from '@open-formulieren/types';
-import {Translation} from '@open-formulieren/types/lib/formio/i18n';
+import {css} from '@emotion/css';
+import {OFExtensions} from '@open-formulieren/types';
+import clsx from 'clsx';
 import {useFormikContext} from 'formik';
-import get from 'lodash.get';
-import {useContext, useEffect, useRef} from 'react';
-import {FormattedMessage} from 'react-intl';
+import {useContext, useState} from 'react';
+import {FormattedMessage, useIntl} from 'react-intl';
 
 import {BuilderContext} from '@/context';
+import {FilterByValueType} from '@/types';
 import {AnyComponentSchema} from '@/types/schemas';
 
-import {DataGrid, DataGridRow, Tab, TabList, TabPanel, Tabs, TextField} from '..//formio';
+import {Component, TextField} from '../formio';
+import ComponentLabel from '../formio/component-label';
 
-export type SchemaKey<S> = keyof S & string;
+type ExtractTranslatableProperties<T> = T extends OFExtensions<infer TK> ? TK : never;
+type StringValueProperties<S> = S extends AnyComponentSchema
+  ? keyof FilterByValueType<S, string | undefined>
+  : never;
```

Comment:
```suggestion
type TranslatableProperties<T> = T extends OFExtensions<infer TK> ? TK : never;
type StringValueProperties<S> = S extends AnyComponentSchema
  ? keyof FilterByValueType<S, string | undefined>
  : never;
```

If I understand correctly, of some FormIO dict T/S these represent the keys that are translatable/Optional[str] and they can be inferred to the bottom type `never`, to make sure dicts that have nu such keys, don't typecheck.

The StringValueProperties I can follow.

But the TranslatableProperties look like they depend on the same type variable name `TK` used in `{openForms?: { translations: ComponentTranslations<TK>}, ...}`. But they're different TK's. And since all keys in `OFExtensions` are marked optional, I don't understand how this type works.

Reply:
Yeah so `S` is essentially "Formio **S**chema". `T` has a similar target, but it's not bound to be assignable to `AnyComponentSchema` because not all component types support translations (columns for example is purely layout, so a generic type here would break the entire component and I prefer building everything against `AnyComponentSchema`).

I prefer the name `ExtractTranslatableProperties` because it does exactly that: figure out which properties are translatable from the given (concrete) schema. E.g.:

```ts
type F = ExtractTranslatableProperties<TextFieldComponentSchema>; // 'label' | 'description' | 'tooltip' | 'defaultValue' | 'placeholder';
type G = ExtractTranslatableProperties<ContentComponentSchema>; // 'html' ;
```

When passing in the union `AnyComponentSchema` you will indeed get a union of all possible translatable properties, but that's not how these components/types are supposed to be used - when you use the component, you need to pass in the concrete schema to really benefit from the type checking.

In the type definitions for each schema, we explicitly pass in which keys can be translated, see for example: https://github.com/open-formulieren/types/blob/b7e6931fbe7fd0d8abe773dde1f69ea82cf9296b/src/formio/components/textfield.ts#L4

> and they can be inferred to the bottom type never, to make sure dicts that have nu such keys, don't typecheck.

Yes, I think? An example of this would be the `columns` formio component type - nothing is user-facing there, so nothing is translatable.

> But the TranslatableProperties look like they depend on the same type variable name TK used in `{openForms?: { translations: ComponentTranslations<TK>}, ...}`.

The optional keys inside aren't relevant - remember that the `extends` meaning is "can be assigned to" and the optional/`undefined` keys make that assignability a little bit easier.

What's then happening, for example with `TextFieldComponentSchema`, is that you see that that component is created via the [`InputComponentSchema`](https://github.com/open-formulieren/types/blob/b7e6931fbe7fd0d8abe773dde1f69ea82cf9296b/src/formio/components/textfield.ts#L6) primitive, and that itself [extends](https://github.com/open-formulieren/types/blob/b7e6931fbe7fd0d8abe773dde1f69ea82cf9296b/src/formio/base.ts#L148) `OFExtensions<TK>`. So, typescript is able to `infer TK` because it sees `OFExtensions<'label' | 'description' | 'tooltip' | 'defaultValue' | 'placeholder'>`, meaning it infers `TK` to be `'label' | 'description' | 'tooltip' | 'defaultValue' | 'placeholder'`

If you lift this up one level to make it more generic, it distributes this logic over all union members: `TextFieldComponentSchema`, `NumberComponentSchema`... to get the full union for `AnyComponentSchema`. I found this to be the only way to achieve this (with the conditional type) because `keyof AnyComponentSchema['openForms']['translations'][SupportedLocales]` seems to express the same, excep that this takes the intersection of all union members which ends up with `never` (because ultimately nothing intersects with the `html` property from the content component).